server/aws-lsp-identity/src/language-server/profiles/profileService.ts

```diff
@@ -0,0 +1,245 @@
+    save(data: ProfileData, init?: SharedConfigInit): Promise<void>
+}
+
+export const ProfileFields = {
+    region: 'region',
+    sso_account_id: 'sso_account_id',
+    sso_role_name: 'sso_role_name',
+    sso_session: 'sso_session',
+} as const
+
+export const SsoSessionFields = {
+    sso_region: 'sso_region',
+    sso_registration_scopes: 'sso_registration_scopes',
+    sso_start_url: 'sso_start_url',
+} as const
+
+export const profileDuckTypers = {
+    SsoTokenProfile: new DuckTyper()
+        .requireProperty(ProfileFields.sso_session)
+        .disallowProperty(ProfileFields.sso_account_id)
+        .disallowProperty(ProfileFields.sso_role_name),
+}
+
+export const ssoSessionDuckTyper = new DuckTyper()
+    .requireProperty(SsoSessionFields.sso_start_url)
+    .requireProperty(SsoSessionFields.sso_region)
+    .optionalProperty(SsoSessionFields.sso_registration_scopes)
+
+export function detectProfileKind(profile: object): ProfileKind {
+    return (profileDuckTypers.SsoTokenProfile.eval(profile) && ProfileKind.SsoTokenProfile) || ProfileKind.Unknown
+}
+
+export function normalizeSettingList(
+    list: string | string[] | null | undefined,
+    delimiter: string = ','
+): string[] | undefined {
+    if (!list) {
+        return undefined
+    }
+
+    if (!Array.isArray(list)) {
+        list = list.split(delimiter)
+    }
+
+    return [...new Set(list)]
+        .map(item => item.trim())
+        .filter(item => item !== '')
+        .sort()
+}
+
+export class ProfileService {
+    constructor(private store: ProfileStore) {}
+
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    async listProfiles(params: ListProfilesParams, token?: CancellationToken): Promise<ListProfilesResult> {
+        // Currently only returns non-legacy sso-session profiles, will return more profile types in the future
+        return await this.tryAsync(
+            this.store.load,
+            messageOrError =>
+                new ListProfilesError(messageOrError, { awsErrorCode: AwsErrorCodes.E_CANNOT_READ_SHARED_CONFIG })
+        )
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    async updateProfile(params: UpdateProfileParams, _token?: CancellationToken): Promise<UpdateProfileResult> {
+        // Currently only supports non-legacy SSO profiles with sso-sessions
+        const result: UpdateProfileResult = {}
+        const options = { ...updateProfileOptionsDefaults, ...params.options }
+
+        // Validate params
+        this.throwOnInvalidProfile(!params.profile, 'Profile required.')
+        const profile = params.profile!
+        this.throwOnInvalidProfile(
+            profile.kind !== ProfileKind.SsoTokenProfile,
+            'Profile must be non-legacy sso-session type.'
+        )
+        this.throwOnInvalidProfile(!profile.name, 'Profile name required.')
+        this.throwOnInvalidProfile(!profile.settings.sso_session, 'Sso-session name required on profile.')
```

Comment:
Would this throw if `profile.settings` is undefined?

Reply:
Also same below at with `!ssoSession.settings`

Reply:
No as the DuckTyper upstream would've failed and the flow would've never made it this far.  However, for defense in-depth and the possibility that this code is used in a different flow in the future, adding a check and unit tests for it.